### PR TITLE
Support an ordered set of options when asserting statsd calls

### PR DESCRIPTION
**Pull request by surrahman**
Currently, when we assert metrics recorded from statsd calls, we allow developers to specify a single set of options for metrics. This prevents us from validating cases where we have multiple metrics with the same name and type being recorded, but with different options, e.g., the following:
```
StatsD.increment('counter', tags: { foo: 1 })
StatsD.increment('counter', tags: { foo: 1 })
StatsD.increment('counter', tags: { foo: 2 })
```
This change allows us to specify an ordered array of metric options via `StatsD::Instrument::Assertions#assert_statsd_calls` so the above case could be validated as shown below:

```
foo_1_metric = StatsD::Instrument::MetricExpectation.new(type: :c, name: 'counter', times: 2, tags: ['foo:1'])
foo_2_metric = StatsD::Instrument::MetricExpectation.new(type: :c, name: 'counter', times: 1, tags: ['foo:2'])
@test_case.assert_statsd_calls([foo_1_metric, foo_2_metric]) do
  StatsD.increment('counter', tags: { foo: 1 })
  StatsD.increment('counter', tags: { foo: 1 })
  StatsD.increment('counter', tags: { foo: 2 })
end
```

### CODE REVIEW

**Comment by wvanbergen** [comment]
Not sure if I get this. Can you share a code snippet of how this would be used?

**Comment by wvanbergen** [comment]
Also, this really needs tests to ensure it works, and stays working.

**Comment by surrahman** [comment]
Oh wow @wvanbergen, you picked up on this quickly! Just using this to share an idea with @fw42 .. it would need tests for sure .. we're looking at an alternate approach now so this might not be needed at all.

File: test/assertions_test.rb [hunk 1]
```diff
@@ -87,13 +87,43 @@ def test_assert_statsd_call
...
 
+    assert_no_assertion_triggered do
+      foo_1_metric_tags = { times: 2, tags: ['foo:1'] }
+      foo_2_metric_tags = { times: 1, tags: ['foo:2'] }
+      @test_case.assert_statsd_increments('counter', [foo_1_metric_tags, foo_2_metric_tags]) do
+        StatsD.increment('counter', tags: { foo: 1 })
+        StatsD.increment('counter', tags: { foo: 1 })
+        StatsD.increment('counter', tags: { foo: 2 })
+      end
+    end
```
Comment (surrahman) [inline test/assertions_test.rb]: @fw42, @wvanbergen, this is the sort of thing I'd like to add support for .. does it make sense to you guys? If there's a better/easier way to do this with what we already have, happy go to that route instead.

File: lib/statsd/instrument/assertions.rb [hunk 2]
```diff
@@ -11,6 +11,10 @@ def assert_statsd_increment(metric_name, options = {}, &block)
     assert_statsd_call(:c, metric_name, options, &block)
   end
 
+  def assert_statsd_increments(metric_name, ordered_metric_options = [], &block)
```
Comment (fw42) [inline lib/statsd/instrument/assertions.rb]: Not sure if adding a new method is the best choice for the API here... Maybe it's more intuitive to change the old API (and support both the old "syntax" and the new one)?

File: lib/statsd/instrument/assertions.rb [hunk 3]
```diff
@@ -11,6 +11,10 @@ def assert_statsd_increment(metric_name, options = {}, &block)
     assert_statsd_call(:c, metric_name, options, &block)
```
Comment (fw42) [inline lib/statsd/instrument/assertions.rb]: Not sure, but if we want to go with the new API, then maybe this old method should now just call the new one (with `ordered_metric_options = [ options ]`?)

File: lib/statsd/instrument/assertions.rb [hunk 4]
```diff
@@ -33,39 +37,56 @@ def assert_statsd_key_value(metric_name, options = {}, &block)
...
 
-  def assert_statsd_call(metric_type, metric_name, options = {}, &block)
-    options[:times] ||= 1
+  def assert_statsd_calls(metric_type, metric_name, ordered_metric_options = [], &block)
     metrics = capture_statsd_calls(&block)
     metrics = metrics.select { |m| m.type == metric_type && m.name == metric_name }
     assert metrics.length > 0, "No StatsD calls for metric #{metric_name} were made."
-    assert options[:times] === metrics.length, "The amount of StatsD calls for metric #{metric_name} was unexpected. Expected #{options[:times].inspect}, found #{metrics.length}"
+    expected_calls_count = 0
+    current_metric_index = 0
```
Comment (fw42) [inline lib/statsd/instrument/assertions.rb]: You can use `each_with_index` instead of `each` and avoid having this variable declared here like this

Comment (fw42) [inline lib/statsd/instrument/assertions.rb]: actually I'm wrong, was misunderstanding your code. ignore. sorry.

**Comment by fw42** [comment]
I think I understand this better now and I agree that it would be useful (and I don't see how we can achieve something similar without this addition). Code looks good to me too, but I think the PR description needs to be a bit clearer (maybe add an example).

tl;dr 👍 🚢 💨

File: test/assertions_test.rb (hunk 1 above)
Comment (wvanbergen) [inline test/assertions_test.rb]: Is the order important? Or do you just want to assert that all the listed configurations occur at least once? Or exactly once?

File: test/assertions_test.rb [hunk 5]
```diff
@@ -87,13 +87,43 @@ def test_assert_statsd_call
...
+    assert_assertion_triggered do
+      foo_1_metric_tags = { times: 2, tags: ['foo:1'] }
+      foo_2_metric_tags = { times: 1, tags: ['foo:2'] }
+      @test_case.assert_statsd_increments('counter', [foo_1_metric_tags, foo_2_metric_tags]) do
+        StatsD.increment('counter', tags: { foo: 1 })
+        StatsD.increment('counter', tags: { foo: 1 })
+        StatsD.increment('counter', tags: { foo: 2 })
+        StatsD.increment('counter', tags: { foo: 2 })
+      end
+    end
```
Comment (wvanbergen) [inline test/assertions_test.rb]: Instead of adding more and more assertions to this test case, I rather add a separate one (or multiple) with descriptions that tell me exactly the behavior that is being tested..

File: test/assertions_test.rb (hunk 1 above)
Comment (surrahman) [inline test/assertions_test.rb]: I think thats a question for whoever is writing the test .. they may want to test that metrics pop up in a certain order .. or they might not care. I haven't seen any tests so far that care about ordering so far. Implementing it with ordering was easy. With respect to number of occurrences, I figured I'd just extend the approach that was already in place .. allow devs to specify the `times` a metric should be seen .. rather than forcing them to repeat asserted metric option sets.

File: lib/statsd/instrument/assertions.rb (hunk 4 above)
Comment (surrahman) [inline lib/statsd/instrument/assertions.rb]: I actually started with `each_with_index` .. just because I wanted to use it :p

File: test/assertions_test.rb (hunk 5 above)
Comment (surrahman) [inline test/assertions_test.rb]: Sure, I'll pull these out into another test case called `test_assert_multiple_statsd_calls`.

Comment (surrahman) [inline test/assertions_test.rb]: @wvanbergen do you think these tests would suffice?

**Comment by Amos47** [comment]
Ping @wvanbergen. Can we merge this? I could really use it for testing multiple statsD calls with different tags. We can override `times` so it works, but it would be cool to be able to assert the tags without having to add another surrounding block.

**Comment by wvanbergen** [comment]
I think this is a good idea, but I don't really like the implementation. 
- This treats metric names as a special case compared to the other metric attributes
- It feels "bolted on" at the wrong level of abstraction, and doesn't feel like a coherent whole. This makes the API not very consistent and hard to learn. What we need is some kind of expectation object.

I would prefer a (low-level) API like this:

``` ruby
expected_calls = [
  StatsD::MetricExpectation.new(name: 'name', tags: [...])
  StatsD::MetricExpectation.new(name: 'other_name', sample_rate: 1.0)
]

assert_statsd_calls(expected_calls, match: :all / :any) do # default = all
  # ...
end
```

Then all the current assertion methods can be implemented on top of this mechanism.


Does this make sense? Or is it over-engineering?

**Comment by Amos47** [comment]
@wvanbergen I think that makes sense.

The issue right now I think is that we have an assert inside the each we have an assert https://github.com/Shopify/statsd-instrument/pull/86/files#diff-1cdd21b1914856daa678effbadef7645L63. This means we can never assert the second call's tags aside from asserting that it was called with `times`.

**Comment by fw42** [comment]
I think refactoring the implementation makes sense, as long as the external API of the existing assertions doesn't change (we don't want to go through our codebase and change 500 users of those assertion helpers).

**Comment by surrahman** [comment]
Working on refactoring this .. @wvanbergen, is the `match: :all / :any ` really necessary? I think all the tests we have operate under the `all` requirement .. allowing `any` might also result in some metrics disappearing without our noticing (i.e., incorrect usage of the `any` option).

**Comment by wvanbergen** [comment]
I'm not sure. If you can implement the current functionality with just `all`, let's start there. we can always add `any` later.

**Comment by surrahman** [comment]
Cool .. I'll start with `all` and we can add in `any` later, as/when needed.

**Comment by surrahman** [comment]
@wvanbergen, is this closer to what you had in mind?

File: lib/statsd/instrument/metric_expectation.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  # @times times [Numeric] :times The number of times this metric is expected
+  # @@ignore_tags @ignore_tags [Array<String>, Hash<String, String>, nil] :@ignore_tags The tags
+  #   to ignore for this metric.
+  def initialize(options = {})
+    super(options)
+    @times = options[:times] or raise ArgumentError, "Metric :times is required."
+    @ignore_tags = StatsD::Instrument::Metric.normalize_tags(options[:ignore_tags])
+  end
+
+  # @private
```
Comment (fw42) [inline lib/statsd/instrument/metric_expectation.rb]: what is this stuff?

File: lib/statsd/instrument/metric_expectation.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
+class StatsD::Instrument::MetricExpectation < StatsD::Instrument::Metric
+
+  attr_accessor :times, :ignore_tags
```
Comment (fw42) [inline lib/statsd/instrument/metric_expectation.rb]: looks like you are never using the writer, so `attr_reader` should be enough?

File: lib/statsd/instrument/metric_expectation.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+  # Initializes a new MetricExpectation instance.
```
Comment (fw42) [inline lib/statsd/instrument/metric_expectation.rb]: I'd say that's pretty obvious :) we usually don't add boilerplate comments like this

File: lib/statsd/instrument/metric_expectation.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  #
+  # @times times [Numeric] :times The number of times this metric is expected
+  # @@ignore_tags @ignore_tags [Array<String>, Hash<String, String>, nil] :@ignore_tags The tags
+  #   to ignore for this metric.
+  def initialize(options = {})
+    super(options)
+    @times = options[:times] or raise ArgumentError, "Metric :times is required."
```
Comment (fw42) [inline lib/statsd/instrument/metric_expectation.rb]: hm this is conventionally done as `@times = options.fetch(:times)` (which raises if the key doesn't exist)

File: lib/statsd/instrument/metric_expectation.rb (hunk 6 above)
Comment (surrahman) [inline lib/statsd/instrument/metric_expectation.rb]: Not needed, will remove

File: lib/statsd/instrument/metric_expectation.rb (hunk 7 above)
Comment (surrahman) [inline lib/statsd/instrument/metric_expectation.rb]: Yep!

File: lib/statsd/instrument/metric_expectation.rb (hunk 8 above)
Comment (surrahman) [inline lib/statsd/instrument/metric_expectation.rb]: Removed

File: lib/statsd/instrument/metric_expectation.rb (hunk 9 above)
Comment (surrahman) [inline lib/statsd/instrument/metric_expectation.rb]: Just following convention in `StatsD::Instrument::Metric#initialize` .. but I can change this.

File: test/assertions_test.rb [hunk 10]
```diff
@@ -149,6 +149,38 @@ def test_assert_statsd_call
     end
   end
 
+  def test_assert_multiple_statsd_calls
+    assert_assertion_triggered do
+      foo_1_metric = StatsD::Instrument::MetricExpectation.new(type: :c, name: 'counter', times: 2, tags: ['foo:1'])
+      foo_2_metric = StatsD::Instrument::MetricExpectation.new(type: :c, name: 'counter', times: 1, tags: ['foo:2'])
+      @test_case.assert_statsd_calls([foo_1_metric, foo_2_metric]) do
+        StatsD.increment('counter', tags: { foo: 1 })
+         StatsD.increment('counter', tags: { foo: 2 })
```
Comment (fw42) [inline test/assertions_test.rb]: indentation here is wrong

File: test/assertions_test.rb [hunk 11]
```diff
@@ -149,6 +149,38 @@ def test_assert_statsd_call
...
+      foo_2_metric = StatsD::Instrument::MetricExpectation.new(type: :c, name: 'counter', times: 1, tags: ['foo:2'])
+      @test_case.assert_statsd_calls([foo_1_metric, foo_2_metric]) do
+        StatsD.increment('counter', tags: { foo: 1 })
+        StatsD.increment('counter', tags: { foo: 1 })
+        StatsD.increment('counter', tags: { foo: 2 })
+        StatsD.increment('counter', tags: { foo: 2 })
+      end
+    end
+
+    assert_assertion_triggered do
```
Comment (fw42) [inline test/assertions_test.rb]: why does this pass? you are expecting twice foo:1 and once foo:2 and that's what the block does. So shouldn't this be `assert_no_assertion_triggered` then?

File: test/assertions_test.rb (hunk 1 above)
Comment (fw42) [inline test/assertions_test.rb]: I don't think we ever care what order metrics appear in. At least I can't think of any use cases for that. I wouldn't add support for that unless someone explicitly asks for it.

File: lib/statsd/instrument/assertions.rb [hunk 12]
```diff
@@ -31,41 +31,56 @@ def assert_statsd_key_value(metric_name, options = {}, &block)
...
+  end
 
-        expected_tags = Set.new(StatsD::Instrument::Metric.normalize_tags(options[:tags]))
-        actual_tags = Set.new(metric.tags)
+  private
 
-        if options[:ignore_tags]
-          ignored_tags = Set.new(StatsD::Instrument::Metric.normalize_tags(options[:ignore_tags])) - expected_tags
-          actual_tags -= ignored_tags
+  def assert_metric_matches_expectation(actual_metric, expected_metric)
```
Comment (fw42) [inline lib/statsd/instrument/assertions.rb]: I feel like this method might just as well be an instance method on the expectation class, so you can just say `expected_metric.matches?(actual_metric)` and that returns true or false (and maybe sets some error message or something).

File: lib/statsd/instrument/assertions.rb [hunk 13]
```diff
@@ -31,41 +31,56 @@ def assert_statsd_key_value(metric_name, options = {}, &block)
...
-        end
+    if expected_metric.tags
 
-        assert_equal expected_tags, actual_tags,
-                     "Unexpected StatsD tags for metric #{metric_name}. Expected: #{expected_tags.inspect}, actual: #{actual_tags.inspect}"
+      expected_tags = Set.new(StatsD::Instrument::Metric.normalize_tags(expected_metric.tags))
+      actual_tags = Set.new(actual_metric.tags)
+
+      if expected_metric.ignore_tags
+        ignored_tags = Set.new(StatsD::Instrument::Metric.normalize_tags(expected_metric.ignore_tags)) - expected_tags
```
Comment (fw42) [inline lib/statsd/instrument/assertions.rb]: `expected_metric.ignore_tags` is already normalized. Why normalize it again?

File: lib/statsd/instrument/assertions.rb [hunk 14]
```diff
@@ -31,41 +31,56 @@ def assert_statsd_key_value(metric_name, options = {}, &block)
...
 
-          if options[:ignore_tags].is_a?(Array)
-            actual_tags.delete_if{ |key| options[:ignore_tags].include?(key.split(":").first) }
-          end
-        end
+    if expected_metric.tags
 
-        assert_equal expected_tags, actual_tags,
-                     "Unexpected StatsD tags for metric #{metric_name}. Expected: #{expected_tags.inspect}, actual: #{actual_tags.inspect}"
+      expected_tags = Set.new(StatsD::Instrument::Metric.normalize_tags(expected_metric.tags))
```
Comment (fw42) [inline lib/statsd/instrument/assertions.rb]: I think `expected_metric.tags` (i.e. the `tags` instance method) could take care of the normalization and just return already normalized tags.

File: lib/statsd/instrument/assertions.rb [hunk 15]
```diff
@@ -31,41 +31,56 @@ def assert_statsd_key_value(metric_name, options = {}, &block)
...
+      actual_tags = Set.new(actual_metric.tags)
+
+      if expected_metric.ignore_tags
+        ignored_tags = Set.new(StatsD::Instrument::Metric.normalize_tags(expected_metric.ignore_tags)) - expected_tags
+        actual_tags -= ignored_tags
+
+        if expected_metric.ignore_tags.is_a?(Array)
```
Comment (fw42) [inline lib/statsd/instrument/assertions.rb]: doesn't `normalize_tags` always return an array (I'm not sure)? if so then this will never be false

Comment (fw42) [inline lib/statsd/instrument/assertions.rb]: actually this is your change. In the original code, this was comparing to the unnormalized list of tags and now you are comparing it to the normalized version. Is that intended? Does that keep the original behaviour?

File: lib/statsd/instrument/assertions.rb [hunk 16]
```diff
@@ -31,41 +31,56 @@ def assert_statsd_key_value(metric_name, options = {}, &block)
...
 
-      metric
+      assert_equal expected_tags, actual_tags,
+        "Unexpected StatsD tags for metric #{actual_metric.name}. Expected: #{expected_tags.inspect}, actual: #{actual_tags.inspect}"
     end
+
+    actual_metric
+  end
+
+  def assert_statsd_call(metric_type, metric_name, options = {}, &block)
```
Comment (fw42) [inline lib/statsd/instrument/assertions.rb]: Why is `options` a parameter? you are never passing that in and all the code always uses the default of `{}`. I think in that case it shouldn't be a parameter. Why would you want to (or need to) pass in options to the MetricExpectation class? All the possible options are overwritten by you.

File: lib/statsd/instrument/metric_expectation.rb [hunk 17]
```diff
@@ -0,0 +1,30 @@
+class StatsD::Instrument::MetricExpectation < StatsD::Instrument::Metric
+
+  attr_accessor :times, :ignore_tags
+
+  # Initializes a new MetricExpectation instance.
+  #
+  # @times times [Numeric] :times The number of times this metric is expected
+  # @@ignore_tags @ignore_tags [Array<String>, Hash<String, String>, nil] :@ignore_tags The tags
+  #   to ignore for this metric.
+  def initialize(options = {})
```
Comment (fw42) [inline lib/statsd/instrument/metric_expectation.rb]: Does the rest of this codebase use options hashes? Because if not then I would suggest using keyword arguments instead.

File: test/assertions_test.rb (hunk 11 above)
Comment (surrahman) [inline test/assertions_test.rb]: Yeah that shouldn't pass .. I'll investigate ..

File: test/assertions_test.rb (hunk 10 above)
Comment (surrahman) [inline test/assertions_test.rb]: Will correct

File: lib/statsd/instrument/assertions.rb (hunk 12 above)
Comment (surrahman) [inline lib/statsd/instrument/assertions.rb]: Will refector ..

File: lib/statsd/instrument/assertions.rb (hunk 13 above)
Comment (surrahman) [inline lib/statsd/instrument/assertions.rb]: Will remove ..

File: lib/statsd/instrument/assertions.rb (hunk 14 above)
Comment (surrahman) [inline lib/statsd/instrument/assertions.rb]: Yep .. will change ..

Comment (surrahman) [inline lib/statsd/instrument/assertions.rb]: Looks like we store tags in "canonical" form (normalized) in the `initialize` method when we create new `StatsD::Instrument::Metric` instances, so maybe we don't need to normalize again?

File: lib/statsd/instrument/assertions.rb (hunk 16 above)
Comment (surrahman) [inline lib/statsd/instrument/assertions.rb]: This was the old declaration that I've left unchanged .. I do fill in `name`, `type` and `times`, but not `value` and `sample_rate` or `tags` .. these are then used when comparing recorded metrics against expected metrics in `StatsD::Instrument::MetricExpectation#matches`.

File: lib/statsd/instrument/metric_expectation.rb (hunk 17 above)
Comment (surrahman) [inline lib/statsd/instrument/metric_expectation.rb]: Yeah, I just followed the pattern in `StatsD::Instrument::Metric#initialize` ..

File: test/assertions_test.rb (hunk 1 above)
Comment (surrahman) [inline test/assertions_test.rb]: Agreed ..

**Comment by wvanbergen** [comment]
If you can test this branch on Shopify core, I am happy to merge this.

File: lib/statsd/instrument/metric_expectation.rb [hunk 18]
```diff
@@ -29,7 +29,7 @@ def matches(actual_metric)
         end
       end
 
-      return expected_tags == actual_tags
+      return expected_tags.subset?(actual_tags)
```
Comment (surrahman) [inline lib/statsd/instrument/metric_expectation.rb]: @wvanbergen, I think this is something you should look at. I'm suggesting that the tags we place in our expectation should at least be a subset of the tags that we see in actual metrics generated by our code .. this is based on the usage I've seen in some of the Shopify tests, whereby developers seem to add assertions and only check for the tags they're interested in. What we had here previously was an exact match check .. which is also correct, just more strict. What would you prefer?

File: test/assertions_test.rb [hunk 19]
```diff
@@ -124,7 +124,7 @@ def test_assert_statsd_call
       end
     end
 
-    assert_assertion_triggered do
+    assert_no_assertion_triggered do
```
Comment (surrahman) [inline test/assertions_test.rb]: This test change is based on the other change in this commit noted above, i.e. the check for tags is less strict that it used to be in that it'll validate that expected tags are a subset of actual tags for each matching metric.

File: lib/statsd/instrument/metric_expectation.rb (hunk 18 above)
Comment (surrahman) [inline lib/statsd/instrument/metric_expectation.rb]: @fw42 does this make sense to you?

File: test/assertions_test.rb (hunk 19 above)
Comment (wvanbergen) [inline test/assertions_test.rb]: I think a subset makes sense. :+1:

If you end up sending more tags, DD always aggregates over them by default anyway. So the presence of additional tags should not break your monitors or dashboard by sending additional tags.

Comment (fw42) [inline test/assertions_test.rb]: one argument against it is that you can now no longer assert on the absence of tags.. Like maybe there is a test that says "if a, then this tag is present, and if not a, then this tag is not present"...

Comment (wvanbergen) [inline test/assertions_test.rb]: I think the simple syntax should cater for the most common use case.

If we want to support more advanced use cases, we can make the `MetricExpectation` class have a more advanced "matcher" API (e.g. for assertion that a tag is absent). Users can construct instances an send them to `assert_statsd_calls`. Let's only cross that bridge when there's a need for it though.

Comment (surrahman) [inline test/assertions_test.rb]: 👍 I've marked it as `# @private`, as discussed on Slack.

**Comment by wvanbergen** [comment]
We've opted to tag `MetricExpectation`, and `assert_statsd_calls` as private, because I am not ready yet to commit to not breaking the API of those for now.

**Comment by Goldenson** [comment]
## Note
FYI, this PR breaks the implementation of `assert_statsd_increment` when dealing with `StatsD.prefix`.

## Example

Setting `StatsD.prefix = 'WhateverPrefix'`

🔴  This is not working anymore:
```
assert_statsd_increment('WhateverPrefix.message') do
  ...
end
```

✅ This is working:
```
assert_statsd_increment('message') do
  ...
end
```

Some folks fixed their tests by removing the prefix in this [PR](https://github.com/Shopify/remember-me/pull/107).

I'm not sure what's the expected behaviour here, just want to report it.

cc @christhomson @wvanbergen
